File: pkg/client/publish.go
package client

import (
	"context"
	"encoding/json"
	"fmt"

	"github.com/love-lena/sextant-initial/pkg/sextantproto"
)

// Publish sends env on subject via core NATS (not JetStream — events
// are persisted by the configured JetStream stream subscription on the
// subject). The envelope is validated before publish so a missing
// TraceID or invalid Kind fails on the publisher, not on every
// downstream consumer.
//
// Publish does not wait for an ack. For RPC, use RPC.
func (c *Client) Publish(ctx context.Context, subject string, env sextantproto.Envelope) error {
	if c.isClosed() {
		return ErrClosed
	}
	if subject == "" {
		return fmt.Errorf("client: Publish requires a non-empty subject")
	}
	if env.Ts.IsZero() {
		env.Ts = sextantproto.NowTimestamp()
	}
	if env.ProtoVersion == "" {
		env.ProtoVersion = sextantproto.ProtoVersion
	}
	if err := env.Validate(); err != nil {
		return fmt.Errorf("client: Publish: %w", err)
	}
	raw, err := json.Marshal(env)
	if err != nil {
		return fmt.Errorf("client: marshal envelope: %w", err)
	}
	if err := ctx.Err(); err != nil {
		return err
	}
	if err := c.nc.Publish(subject, raw); err != nil {
		return fmt.Errorf("client: publish %s: %w", subject, err)
	}
	return nil
}

File: clients/typescript/src/publish.ts
/**
 * Publish implementation. Mirrors pkg/client/publish.go.
 *
 * Envelope is validated before publish so a missing trace_id or
 * invalid kind fails on the publisher, not on every downstream
 * consumer. The publish goes via core NATS (not JetStream) — events
 * are persisted by the JetStream stream subscription on the subject.
 */

import type { Client } from "./client.js";
import { encodeEnvelope, validateEnvelope, formatTimestamp } from "./envelope.js";
import { PROTO_VERSION } from "./proto_version.js";
import type { Envelope } from "./types.generated.js";

export async function publish(client: Client, subject: string, env: Envelope): Promise<void> {
  client.ensureOpen();
  if (!subject) throw new Error("client: publish requires a non-empty subject");

  if (!env.ts) env.ts = formatTimestamp();
  if (!env.proto_version) env.proto_version = PROTO_VERSION;

  validateEnvelope(env);
  const bytes = encodeEnvelope(env);
  client.nc.publish(subject, bytes);
  // Flush so the publish is on the wire before we return — matches
  // the spec's request/reply ordering expectations for the RPC path.
  await client.nc.flush();
}
